core/src/main/java/io/questdb/std/MmapCache.java

```diff
@@ -0,0 +1,276 @@
+            mmapFileCache.putAt(fdMapIndex, fileCacheKey, record);
+
+            // Point the returned address to the correct offset
+            mmapAddrCache.put(address, record);
+
+            return address;
+        }
+    }
+
+    /**
+     * Returns number of times cached memory mappings were reused.
+     */
+    public long getReuseCount() {
+        return mmapReuseCount;
+    }
+
+    /**
+     * Checks if memory mapping has only one active reference.
+     */
+    public boolean isSingleUse(long address) {
+        var cacheRecord = mmapAddrCache.get(address);
+        return cacheRecord != null && cacheRecord.count == 1;
+    }
+
+    /**
+     * Resizes existing memory mapping, reusing or creating new mapping as needed.
+     */
+    public long mremap(int fd, long fileCacheKey, long address, long previousSize, long newSize, long offset, int flags, int memoryTag) {
+        // TODO: handle the offset
+        if (offset != 0 || fileCacheKey == 0 || !Files.FS_CACHE_ENABLED) {
+            return mremap0(fd, address, previousSize, newSize, offset, flags, memoryTag, memoryTag);
+        }
+
+        long unmapPtr = 0, unmapLen = 0;
+        int unmapTag = 0;
+        long newAddress = 0;
+
+        synchronized (this) {
+
+            int addrMapIndex = mmapAddrCache.keyIndex(address);
+            assert addrMapIndex < 0 : "old address is not found in mmap cache";
+
+            MmapCacheRecord record = mmapAddrCache.valueAt(addrMapIndex);
+
+            if (newSize >= previousSize) {
+                if (record.length >= newSize) {
+                    // Address is already long enough, just return it
+                    return address;
+                }
+
+                // Check if someone else remapped this to a larger size
+                int fdIndex = mmapFileCache.keyIndex(fileCacheKey);
+                if (fdIndex < 0) {
+                    MmapCacheRecord updatedCacheRecord = mmapFileCache.valueAt(fdIndex);
+                    if (updatedCacheRecord.length >= newSize) {
+                        // Cache for the FD is updated by someone else
+                        // The fd cache record is already long enough, just return the address
+                        updatedCacheRecord.count++;
+                        newAddress = updatedCacheRecord.address;
+                        mmapReuseCount++;
+
+                        record.count--;
+                        if (record.count == 0) {
+                            // The old cache record is not used anymore
+                            mmapAddrCache.removeAt(addrMapIndex);
+                            unmapPtr = record.address;
+                            unmapLen = record.length;
+                            unmapTag = record.memoryTag;
+                            record.address = 0;
+                        }
+                    }
+                }
+            }
+
+            if (newAddress == 0) {
```

Comment:
there is a possibility of ref counting to go to bin, if "updatedCacheRecord.address == 0", count could be decremented twice. More robust logic is needed here

Reply:
We cannot have a cached record with address == 0. I added an assertion with comments
```
                        // We should not store zero addresses in the cache. We do not cache if someone maps 0 length,
                        // and we do not allow to remap to 0 length.
                        assert newAddress != 0;
```